### PR TITLE
Rare interests core project

### PR DESCRIPTION
**Pull request by BoominHutJr**
Finished basic implementation, working on test cases.

### CODE REVIEW

File: pipeline/matching/generators/rare_interests.py [hunk 1]
```diff
@@ -1,7 +1,10 @@
-from typing import Iterator
+# used to count occurences in a list
```
Comment (erikqu) [inline pipeline/matching/generators/rare_interests.py:1]: Remove misc. comments unless clever logic (applies everywhere)

File: pipeline/matching/generators/rare_interests.py [hunk 2]
```diff
@@ -14,6 +17,78 @@ def __init__(self, max_frequency: float = 0):
...
+                    metadata=MatchMetadata(
+                        generator="rareInterestsGenerator",
+                        score=round(rarityScore, 2),
+                        rareInterests=rareInterests_,
+                    ),
+                )
+            ]
+            yield from m
+
+        usrList = inp.users
```
Comment (erikqu) [inline pipeline/matching/generators/rare_interests.py:43]: feel free to rename the input var name in this method

File: pipeline/matching/generators/rare_interests.py [hunk 3]
```diff
@@ -14,6 +17,78 @@ def __init__(self, max_frequency: float = 0):
...
     def generate(self, inp: MatchingInput) -> Iterator[Match]:
-        # TODO: Implement your generator
-        inp.logger.info("Ayman Fahsi was here!")
-        yield from []
+        # check for null inputs
+        rareInterests_ = []  # type: List[str]
+        rarestInterest = " "
+        rarityScore = 0.0
+        matchNames = set("None")
+        m = []
```
Comment (erikqu) [inline pipeline/matching/generators/rare_interests.py:25]: Lets name this something meaningful

File: pipeline/matching/generators/rare_interests.py [hunk 4]
```diff
@@ -14,6 +17,78 @@ def __init__(self, max_frequency: float = 0):
...
+        # convert to set and cback to list to remove duplicates
+        #     rareInterests_ = set(rareInterests_)
+        #    rareInterests_ = list(rareInterests_)
+
+        m = [
+            Match(
+                users=matchNames,
+                metadata=MatchMetadata(
+                    generator="rareInterestsGenerator",
+                    score=round(rarityScore, 2),
```
Comment (erikqu) [inline pipeline/matching/generators/rare_interests.py:89]: Nice, rounding this was a good idea

**Review by erikqu — CHANGES_REQUESTED** [review]

**Comment by BoominHutJr** [comment]
- added tests to cover cases  in which there are 1 rare interest, multiple rare interests, and all interests are equally rare
- fixed null input check
- fixed rare interest list reflecting rare interests of the community rather than individual matches

**Comment by BoominHutJr** [comment]
- added test for scenario of a community that has no common interests

File: pipeline/matching/generators/rare_interests_test.py [hunk 5]
```diff
@@ -37,13 +37,143 @@ def test_example():
...
+        community="test",
+        release="2022-06-26",
+        users=[
+            User(uid="1", displayName="A", interests=[hiking, swimming]),
+            User(uid="2", displayName="B", interests=[swimming, reading]),
+            User(uid="3", displayName="C", interests=[reading, cooking]),
+            User(uid="4", displayName="D", interests=[swimming, cooking]),
+        ],
+        recent_matches=[],
+        interests=[hiking, swimming, reading],
```
Comment (erikqu) [inline pipeline/matching/generators/rare_interests_test.py:64]: Should we have cooking here?

Comment (BoominHutJr) [reply pipeline/matching/generators/rare_interests_test.py:64]: yes, thanks for catching that

**Comment by bridgettorres** [comment]
LGTM!

File: pipeline/matching/generators/rare_interests.py [hunk 6]
```diff
@@ -13,7 +16,80 @@ def __init__(self, max_frequency: float = 0):
...
+                    rareInterests=rareInterests_,
+                ),
+            )
+        ]
+        yield from resultingMatch
```
Comment (bridgettorres) [inline pipeline/matching/generators/rare_interests.py:95]: Everything looks clean and easy to read, LGTM!

File: pipeline/matching/generators/rare_interests_test.py [hunk 7]
```diff
@@ -37,13 +37,143 @@ def test_example():
...
+            users={"-1", "-2"},
+            metadata=MatchMetadata(
+                generator="rareInterestsGenerator",
+                score=float(-1),
+                rareInterests=[],
+            ),
+        )
+    ]
+    assert len(actual) == 1
+    assert actual == expected
```
Comment (bridgettorres) [inline pipeline/matching/generators/rare_interests_test.py:176]: Good test cases, clear names!

File: pipeline/matching/generators/rare_interests.py [hunk 8]
```diff
@@ -13,7 +16,80 @@ def __init__(self, max_frequency: float = 0):
...
+            for commonInterest in commonInterestList:
+                rarestInterest = min(
+                    rarityScoreDict.items(), key=lambda x: x[1]
+                )[0]
+                if rarestInterest in commonInterestList:
+                    matchNames = {userMatch[0].uid, userMatch[1].uid}
+                    rareInterests_.append(rarestInterest)
+                rarityScore = float(rarityScoreDict[rarestInterest])
+
+        # convert to set and back to list to remove duplicates
```
Comment (erikqu) [inline pipeline/matching/generators/rare_interests.py:81]: Lets remove this comment

File: pipeline/matching/generators/rare_interests.py [hunk 9]
```diff
@@ -13,7 +16,80 @@ def __init__(self, max_frequency: float = 0):
...
+        # generates matches and finds rarest interests and thier score
+        matches = list(it.combinations(usrList, 2))
+        for userMatch in matches:
+            commonInterestList = list(
+                set([x.code for x in userMatch[0].interests])
+                & set([x.code for x in userMatch[1].interests])
+            )
+            for commonInterest in commonInterestList:
+                rarestInterest = min(
+                    rarityScoreDict.items(), key=lambda x: x[1]
```
Comment (erikqu) [inline pipeline/matching/generators/rare_interests.py:74]: This could be perceived as clever, maybe a comment here would be nice

**Review by erikqu — APPROVED** [review]

**Pull request merged**